### DISCUSSION TITLE
Our Onboarding Issues (Continued from #129)

### DISCUSSION THREAD

**Discussion by ianks** [socketry/falcon#131] [General]
Continuing from #129 

---

>     1. What is hard about `falcon serve --bind http://localhost`?

Didn't know that would work actually, thank you!

>     2. Falcon doesn't have a fork model like any other server (which is so complicated that requires hooks, which I think is crazy). Your config.ru will be loaded in the child process _OR_ a thread, so you just put the code you need in that file and that file only.

I think this makes sense. It seems like `preload.rb` handles the use case of `before_fork`, but what about `after_fork`? 

For example, when preloading our app, DB schemas / types are inferred and setup by Sequel. In order to do this, it establishes a DB connection. However, we need to disconnect from the socket after a process is forked. How could I accomplish this with Falcon?

>     3. `guard` in general is pretty much unmaintained at this point. `guard-falcon` is a best effort to support it but obviously as a single developer, it's impossible to support the entire eco-system while maintaining a full time job and a family. If you have specific issues, feel free to contribute back.

Fair enough, it's not a deal-breaker by any means for us. I actually think this may "just work" if I could figure out the correct preloading hooks, etc. I'll give it another shot.

**Comment by ioquatix** [comment]
Thanks for following up.

I agree that we need to clearly state the life cycle guarantees.

Falcon needs to enforce those guarantees, but some types of process restart may cause immediate shutdown.

For `after_fork` you can simply write:

```
# in myapp.rb
... after fork code ...

# in config.ru
require_relative 'myapp`
```

Because `config.ru` will be loaded once per child process, or in the case of threads, once per thread.

`require_relative` will avoid loading the code more than once in the same process. So it depends on what you want.

Can you share more details about what you want to do in your `before_fork` and `after_fork` hooks?

Regarding `guard`, I do want to improve support. But it's low priority right now. However your feedback is helpful.
